### DISCUSSION TITLE
Migrate from `http4s-crypto` to `fs2.hashing`

### DISCUSSION THREAD

**Discussion by danicheg** [http4s/http4s#7508] [Ideas]
Given the introduction of the `fs.hashing` package in `v3.11.0`, I wonder if it makes sense to migrate to it from our self-maintained `http4s-crypto`. Surely, this isn’t achievable in the `0.23` series due to backwards compatibility concerns. But what about the `1.0` series? In my humble opinion, it makes sense to reduce the number of self-maintained modules and migrate to the capabilities provided by our upstream, considering the quality of that software! I quickly checked the API from the fs2 module, and everything is covered except for generating `org.http4s.crypto.HmacKeyGen`. However, if I'm not mistaken, we only use it in a single place, and it's an implementation detail, not part of the public API. WDYT?

**Comment by bplommer** [comment]
I agree in principle, with the proviso that I know nothing specific about those two crypto libraries.

**Comment by armanbilge** [comment]
Overall I agree, I'd be really glad to retire the `http4s-crypto` repository.

Note that `fs2.hashing` cannot be used as a direct replacement for `http4s-crypto`. `fs2.hashing` only supports Node.js but `http4s-crypto` supports Node.js, browsers, and other JS runtimes. Also, `fs2.hashing` fundamentally cannot support browsers because there are no APIs available for incrementally building a hash from a stream of data.

> Surely, this isn’t achievable in the `0.23` series due to backwards compatibility concerns. But what about the `1.0` series?

In https://github.com/http4s/http4s/discussions/7386 we discuss an http4s 1.0 that is significantly stripped down. I'm hopeful that if we go forward with that plan, many of the current uses of crypto routines will "go away" into libraries that can make their own decisions about which platforms to support.
